### PR TITLE
Rewrite import insertion

### PR DESCRIPTION
**Pull request by Veykril**
This is my attempt at refactoring the import insertion #3947. I hope what I created here is somewhat in line with what was requested, it wouldn't surprise me .

`common_prefix` is a copy from `merge_imports.rs` so those should be unified somewhere, `try_merge_trees` is also copied from there but slighly modified to take the `MergeBehaviour` enum into account.
`MergeBehaviour` should in the end become a configuration option, and the order if `ImportGroup` probably as well?

I'm not too familiar with the assist stuff and the like which is why I dont know what i have to do with `insert_use_statement` and `find_insert_use_container` for now.

I will most likely add more test cases in the end as well as I currently only tried to hit every path in `find_insert_position`. 
Some of the merge tests also fail atm due to them not sorting what they insert. There is also this test case I'm not sure if we want to support it. I would assume we want to? https://github.com/rust-analyzer/rust-analyzer/pull/5935/files#diff-6923916dd8bdd2f1ab4b984adacd265fR540-R547

The entire module was rewritten so looking at the the file itself is probably better than looking at the diff.

Regarding the sub issues of #3947:
- #3301: This is fixed with the rewrite, what this implementation does is that it scans through the first occurence of groupings and picks the appropriate one out. This means the user can actually rearrange the groupings on a per file basis to their liking. If a group isnt being found it is inserted according to the `ImportGroup` variant order(Would be nice if this was configurable I imagine).
- #3831: This should be fixed with the introduced `MergeBehaviour` enum and it's `Last` variant.
- #3946: This should also be [fixed](https://github.com/rust-analyzer/rust-analyzer/pull/5935/files#diff-6923916dd8bdd2f1ab4b984adacd265fR87)
- #5795: This is fixed in the sense that the grouping search picks the first group that is of the same kind as the import that is being added. So if  there is a random import in the middle of the program it should only be considered if there is no group of the same kind in the file already present.
- the last point in the list I havent checked yet, tho I got the feeling that it's not gonna be too simple as that will require knowledge of whether in this example `ast` is a crate or the module that is already imported.

### CODE REVIEW

**Comment by kjeremy** [comment]
Does this fix #5913?

**Comment by Veykril** [comment]
It seems to fix it, instead of erroring it turns the import into `use token::TokenKind::{self::*, self};`

File: crates/assists/src/utils/insert_use.rs [hunk 1]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
-        make_assist(&action, &target, builder);
+    algo::insert_children(&where_, insert_position, to_insert)
+}
+
+fn try_merge_imports(
+    old: &ast::Use,
+    new: &ast::Use,
+    merge_behaviour: MergeBehaviour,
+) -> Option<ast::Use> {
+    // dont merge into re-exports
```
Comment (lnicola) [inline crates/assists/src/utils/insert_use.rs:86]: Nit:

```suggestion
    // don't merge into re-exports
```

File: crates/assists/src/utils/insert_use.rs [hunk 2]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
 ) {
-    let target = path_to_import.split("::").map(SmolStr::new).collect::<Vec<_>>();
-    let container = find_insert_use_container(position, ctx);
+    insert_use(position.clone(), make::path_from_text(path_to_import), Some(MergeBehaviour::Full));
+}
+
+pub fn insert_use(
+    where_: SyntaxNode,
+    path: ast::Path,
+    merge_behaviour: Option<MergeBehaviour>,
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:40]:
```suggestion
    merge: Option<MergeBehaviour>,
```

I took me a while to get why this needs to be an `Option`, maybe renaming to `merge` help?

File: crates/assists/src/utils/insert_use.rs [hunk 3]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+            buf.push(make::tokens::single_newline().into());
+        }
+
+        buf.push(use_item.syntax().clone().into());
+
+        if add_blank == AddBlankLine::After {
+            buf.push(make::tokens::single_newline().into());
+        } else if add_blank == AddBlankLine::AfterTwice {
+            buf.push(make::tokens::single_newline().into());
+            buf.push(make::tokens::single_newline().into());
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:72]: `make::tokens::blank_line()` maybe?

File: crates/assists/src/utils/insert_use.rs [hunk 4]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+    let to_insert: Vec<SyntaxElement> = {
+        let mut buf = Vec::new();
+
+        if add_blank == AddBlankLine::Before {
+            buf.push(make::tokens::single_newline().into());
+        }
+
+        buf.push(use_item.syntax().clone().into());
+
+        if add_blank == AddBlankLine::After {
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:68]: Alternatively, `match`

File: crates/assists/src/utils/insert_use.rs [hunk 5]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+    algo::insert_children(&where_, insert_position, to_insert)
+}
+
+fn try_merge_imports(
+    old: &ast::Use,
+    new: &ast::Use,
+    merge_behaviour: MergeBehaviour,
+) -> Option<ast::Use> {
+    // dont merge into re-exports
+    if old.visibility().map(|vis| vis.pub_token()).is_some() {
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:87]: Hm, internal `map` is no-op, no?

File: crates/assists/src/utils/insert_use.rs [hunk 6]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+    let lhs = old.split_prefix(&lhs_prefix);
+    let rhs = new.split_prefix(&rhs_prefix);
+    let lhs_tl = lhs.use_tree_list()?;
+    let rhs_tl = rhs.use_tree_list()?;
+
+    // if we are only allowed to merge the last level check if the split off paths are only one level deep
+    if merge_behaviour == MergeBehaviour::Last && use_tree_list_is_nested(&lhs_tl)
+        || use_tree_list_is_nested(&rhs_tl)
+    {
+        return None;
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:121]: Such special cases could make use of marks: https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/test_utils/src/mark.rs

File: crates/assists/src/utils/insert_use.rs [hunk 7]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
     }
+
+    let should_insert_comma = lhs_tl
+        .r_curly_token()
+        .and_then(|it| skip_trivia_token(it.prev_token()?, Direction::Prev))
+        .map(|it| it.kind() != T![,])
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:127]: Perhaps `!= Some(T![,])`?

File: crates/assists/src/utils/insert_use.rs [hunk 8]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+    let mut to_insert: Vec<SyntaxElement> = Vec::new();
+    if should_insert_comma {
+        to_insert.push(make::token(T![,]).into());
+        to_insert.push(make::tokens::single_space().into());
+    }
+    to_insert.extend(
+        rhs_tl
+            .syntax()
+            .children_with_tokens()
+            .filter(|it| it.kind() != T!['{'] && it.kind() != T!['}']),
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:138]:
```suggestion
            .filter(|it| !matches!(it.kind(), T!['{'] | T!['}'])),
```

File: crates/assists/src/utils/insert_use.rs [hunk 9]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
-                    .syntax()
-                    .ancestors()
-                    .find_map(ast::Use::cast)
-                    .map(|it| it.syntax().clone()),
-                true,
-            );
-        }
-    };
+    #[test]
+    fn insert_middle_pnested() {
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:346]:
```suggestion
    fn insert_middle_nested() {
```

File: crates/assists/src/utils/insert_use.rs [hunk 10]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
-    fn better(left: ImportAction, right: ImportAction) -> ImportAction {
-        if left.is_better(&right) {
-            left
-        } else {
-            right
-        }
+    #[test]
+    fn insert_middle() {
+        check_none(
+            "std::bar::E",
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:316]: Maybe call this and other inserted imports as `::EE` or something like that, such that they better stand out in tests?

File: crates/assists/src/utils/insert_use.rs [hunk 11]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+            r"use std::bar::A;
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:317]: Can we format this with a leading blank line?

```
    r"
use std::bar::A;
```

?

File: crates/assists/src/utils/insert_use.rs [hunk 12]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+        check(path, ra_fixture_before, ra_fixture_after, Some(MergeBehaviour::Full))
+    }
+
+    fn check_last(path: &str, ra_fixture_before: &str, ra_fixture_after: &str) {
+        check(path, ra_fixture_before, ra_fixture_after, Some(MergeBehaviour::Last))
+    }
+
+    fn check_none(path: &str, ra_fixture_before: &str, ra_fixture_after: &str) {
+        check(path, ra_fixture_before, ra_fixture_after, None)
     }
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:582]: Nice tests!

File: crates/assists/src/utils/insert_use.rs [hunk 13]
```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
...
+        )
+    }
+
+    // should this be a thing?
+    #[test]
+    fn split_merge() {
+        check_last(
+            "std::fmt::Result",
+            r"use std::{fmt, io};",
+            r"use std::fmt::Result;
```
Comment (matklad) [inline crates/assists/src/utils/insert_use.rs:544]: Sholdn't this be `use std::fmt::{self, Result}`?

**Review by matklad — COMMENTED** [review]
Yes, this looks exactly how I would imagine it! Regarding `find_insert_use_container` and `insert_use_statement`, I think ideally we should just remove this functions from this module. Nothing here should mention `AssistContext` or `TextEditBuilder`. But it is OK to leave them in temporary, if that eases migration. 

Similar point about code duplication -- it is fine if it make the rewrite easier, removing duplication can be done in a separate PR. 

Basically, I think we should merge this as soon as the tests are green!

Excellent work @Veykril !

**Comment by matklad** [comment]
>There is also this test case I'm not sure if we want to support it. I would assume we want to?

It's OK to not support weird cases initially, if we don't do something completely crazy.

**Comment by matklad** [comment]
For config, I don't think we should handle it in this PR necessary, but it should work like this:

* define

```rust
struct InsertUseConfig { ... }
```

* add it as a field to `AssistConfig`

* add required user-visible options in package.json and `ConfigData`

* thread them to `AssistConfig` in https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/rust-analyzer/src/config.rs#L23

File: crates/assists/src/utils/insert_use.rs (hunk 2 above)
Comment (Veykril) [reply crates/assists/src/utils/insert_use.rs]: That's a good point, figured that the `Option` wrapping might be a bit confusing, changed the name and added a small doc comment.

File: crates/assists/src/utils/insert_use.rs (hunk 3 above)
Comment (Veykril) [reply crates/assists/src/utils/insert_use.rs]: I'm honestly not sure how I would implement that function, I'm not too used to how `rowan` and the like works yet. So I'm unsure how I would fetch two newlines as a single `SyntaxToken`.

File: crates/assists/src/utils/insert_use.rs (hunk 5 above)
Comment (Veykril) [reply crates/assists/src/utils/insert_use.rs]: Ah this was supposed to be `and_then`.

File: crates/assists/src/utils/insert_use.rs (hunk 13 above)
Comment (Veykril) [reply crates/assists/src/utils/insert_use.rs]: Yes, I forgot to change that back as this wasn't working yet. I will mark it as `#[ignore]` for now since I think I wont implement this case in this PR.

File: crates/assists/src/utils/insert_use.rs (hunk 6 above)
Comment (Veykril) [reply crates/assists/src/utils/insert_use.rs:129]: Ah that seems nice, I actually haven't written a test for this case I think haha

**Comment by Veykril** [comment]
Almost all of the current test failures are due to the tests using `check_assist` which panics on an `Option::unwrap` in https://github.com/rust-analyzer/rust-analyzer/blob/7f79dbc84f5d69ec6f48fa8d67bdfa0525c5dd34/crates/assists/src/tests.rs#L84 which I assume comes from me not using the AssistContext/TextEditBuilder anymore.

**Comment by matklad** [comment]
bors r+

:tada:
